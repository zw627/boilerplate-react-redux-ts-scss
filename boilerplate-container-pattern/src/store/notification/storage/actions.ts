// ------------------------------------------------------ //
// Actions

export const SET = "notificationStorage/set";

// ------------------------------------------------------ //
// Action creators

export const set = (visibility: boolean): NotificationStorageActions => ({
  type: "notificationStorage/set",
  payload: {
    visibility,
  },
});

// ------------------------------------------------------ //
// Types - action

interface Set {
  type: typeof SET;
  payload: {
    visibility: boolean;
  };
}

export type NotificationStorageActions = Set | { type: undefined };

// ------------------------------------------------------ //
// Types - state

export interface NotificationStorageState {
  visibility: boolean;
}
