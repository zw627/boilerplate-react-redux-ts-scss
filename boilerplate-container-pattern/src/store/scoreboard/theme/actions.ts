// ------------------------------------------------------ //
// Actions

export const TOGGLE = "theme/toggle";

// ------------------------------------------------------ //
// Action Creators

export const toggle = (): ThemeActions => ({
  type: "theme/toggle",
});

// ------------------------------------------------------ //
// Types - action

interface Toggle {
  type: typeof TOGGLE;
}

export type ThemeActions = Toggle | { type: undefined };

// ------------------------------------------------------ //
// Types - state

export interface ThemeState {
  isLightMode: boolean;
}
