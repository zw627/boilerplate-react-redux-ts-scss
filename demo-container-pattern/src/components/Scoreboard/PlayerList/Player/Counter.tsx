import React from "react";

import getCurrentDate from "Utils/getCurrentDate";
import { CounterProps } from "Containers/Scoreboard/PlayerList/Player/Counter";

const Counter: React.FC<CounterProps> = ({
  id,
  score,
  update,
}: CounterProps) => {
  // Handlers
  function handleScoreIncrement(): void {
    update(id, 1, getCurrentDate());
  }
  function handleScoreDecrement(): void {
    update(id, -1, getCurrentDate());
  }

  // JSX elemetns
  const decrementButton = (
    <button className="counter-action decrement" onClick={handleScoreDecrement}>
      -
    </button>
  );
  const incrementButton = (
    <button className="counter-action increment" onClick={handleScoreIncrement}>
      +
    </button>
  );

  return (
    <div className="counter">
      {decrementButton}
      <div className="counter-score">{score}</div>
      {incrementButton}
    </div>
  );
};

export default Counter;
