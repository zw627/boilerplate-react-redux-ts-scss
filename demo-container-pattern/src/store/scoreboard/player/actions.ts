// ------------------------------------------------------ //
// Actions

export const ADD = "player/add";
export const REMOVE = "player/remove";
export const UPDATE_SCORE = "player/update";
export const SELECT = "player/select";

// ------------------------------------------------------ //
// Action creators

export const add = (id: string, name: string, date: string): PlayerActions => ({
  type: "player/add",
  payload: {
    id,
    name,
    date,
  },
});

export const remove = (id: string): PlayerActions => ({
  type: "player/remove",
  payload: {
    id,
  },
});

export const update = (
  id: string,
  delta: number,
  date: string
): PlayerActions => ({
  type: "player/update",
  payload: {
    id,
    delta,
    date,
  },
});

export const select = (id: string): PlayerActions => ({
  type: "player/select",
  payload: {
    id,
  },
});

// ------------------------------------------------------ //
// Types - action

interface Add {
  type: typeof ADD;
  payload: {
    id: string;
    name: string;
    date: string;
  };
}

interface Remove {
  type: typeof REMOVE;
  payload: {
    id: string;
  };
}

interface Update {
  type: typeof UPDATE_SCORE;
  payload: {
    id: string;
    delta: number;
    date: string;
  };
}

interface Select {
  type: typeof SELECT;
  payload: {
    id: string;
  };
}

export type PlayerActions =
  | Add
  | Remove
  | Update
  | Select
  | { type: undefined };

// ------------------------------------------------------ //
// Types - state

export interface PlayerObject {
  id: string;
  name: string;
  score: number;
  created: string;
  updated: string;
  isSelected: boolean;
}

export interface PlayerState {
  playerList: PlayerObject[];
}
