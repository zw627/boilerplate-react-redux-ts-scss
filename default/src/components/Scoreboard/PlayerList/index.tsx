import React from "react";
import { useSelector } from "react-redux";

import { AppState } from "Store/index";
import Player from "./Player";
import { playerListSelector } from "Utils/selectors";

const PlayerList: React.FC<{}> = () => {
  // State (redux)
  const playerList = useSelector((state: AppState) =>
    playerListSelector(state)
  );

  // JSX elements
  const playerDivs = playerList.map((player, index) => {
    return <Player key={player.id} index={index} />;
  });

  return <div className="players">{playerDivs}</div>;
};

export default PlayerList;
