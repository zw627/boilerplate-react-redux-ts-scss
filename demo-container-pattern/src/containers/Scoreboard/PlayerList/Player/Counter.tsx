import { connect } from "react-redux";

// Component
import Counter from "Components/Scoreboard/PlayerList/Player/Counter";

// Store
import { AppState } from "Store/index";
import { update } from "Store/scoreboard/player/actions";
import { playerListSelector } from "Store/selectors";

interface CounterStateProps {
  id: string;
  score: number;
}

export interface CounterProps extends CounterStateProps {
  update: typeof update;
}

const mapStateToProps = (
  state: AppState,
  ownProps: { index: number }
): CounterStateProps => {
  const { index } = ownProps;
  const { id, score } = playerListSelector(state)[index];
  return { id, score };
};

const actionCreators = { update };

export default connect(mapStateToProps, actionCreators)(Counter);
