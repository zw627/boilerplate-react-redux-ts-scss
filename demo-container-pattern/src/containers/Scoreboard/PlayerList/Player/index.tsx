import { connect } from "react-redux";

// Component
import Player from "Components/Scoreboard/PlayerList/Player";

// Store
import { AppState } from "Store/index";
import { playerListSelector } from "Store/selectors";
import { remove, select } from "Store/scoreboard/player/actions";

interface PlayerStateProps {
  index: number;
  id: string;
  name: string;
  isSelected: boolean;
}

export interface PlayerProps extends PlayerStateProps {
  remove: typeof remove;
  select: typeof select;
}

const mapStateToProps = (
  state: AppState,
  ownProps: { index: number }
): PlayerStateProps => {
  const { index } = ownProps;
  const { id, name, isSelected } = playerListSelector(state)[index];
  return {
    index,
    id,
    name,
    isSelected,
  };
};

const actionCreators = { remove, select };

export default connect(mapStateToProps, actionCreators)(Player);
