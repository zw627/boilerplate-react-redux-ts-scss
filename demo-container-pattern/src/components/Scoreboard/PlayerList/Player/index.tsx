import React from "react";

import Counter from "Containers/Scoreboard/PlayerList/Player/Counter";
import Crown from "Containers/Scoreboard/PlayerList/Player/Crown";
import { PlayerProps } from "Containers/Scoreboard/PlayerList/Player";

const Player: React.FC<PlayerProps> = ({
  index,
  id,
  name,
  isSelected,
  remove,
  select,
}: PlayerProps) => {
  // Handlers
  function handlePlayerRemove(): void {
    remove(id);
  }
  function handlePlayerSelect(): void {
    select(id);
  }

  // JSX elements
  const playerRemoveAnchor = (
    <a
      className={
        isSelected ? "remove-player remove-player-selected" : "remove-player"
      }
      onClick={handlePlayerRemove}
    >
      âœ•
    </a>
  );
  const playerNameDiv = (
    <div className="player-name">
      {playerRemoveAnchor}
      <Crown index={index} />
      <span onClick={handlePlayerSelect}>{name}</span>
    </div>
  );

  return (
    <div className={isSelected ? "player player-selected" : "player"}>
      {playerNameDiv}
      <Counter index={index} />
    </div>
  );
};

export default Player;
