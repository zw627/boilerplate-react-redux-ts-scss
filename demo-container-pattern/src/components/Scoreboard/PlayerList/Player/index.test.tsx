import React from "react";
import { Provider } from "react-redux";
import { Store } from "redux";
import { mount, ReactWrapper } from "enzyme";

import { remove, select } from "Store/scoreboard/player/actions";
import Player from "Components/Scoreboard/PlayerList/Player/index";
import { setupStore, matchMultiLength } from "Utils/testUtils";
import { playerListSelector } from "Store/selectors";

import { AppState } from "Store/index";
import { PlayerProps } from "Containers/Scoreboard/PlayerList/Player";

jest.mock("Store/scoreboard/player/actions");

// Pass down the props from mock store to wrapper
function setupProps(state = setupStore().getState() as AppState): PlayerProps {
  const { id, name, isSelected } = playerListSelector(state)[0];
  return {
    index: 0,
    id,
    name,
    isSelected,
    remove,
    select,
  };
}

function setupWrapper(
  mockStore = setupStore(),
  props = setupProps(mockStore.getState())
): ReactWrapper {
  return mount(
    <Provider store={mockStore}>
      <Player {...props} />
    </Provider>
  ) as ReactWrapper;
}

describe("Scoreboard/Player", () => {
  let mockStore: Store;
  let wrapper: ReactWrapper;

  // Setup wrapper
  beforeEach(() => {
    wrapper = setupWrapper();
  });

  test("compare to the last snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should display player name", () => {
    // Name of the player at index 0, because we passed 0 to setupWrapper
    expect(wrapper.find("span").text()).toEqual("Alpha");
  });

  it("should update related classes if a player is selected", () => {
    // Change `isSelected` of player at index 0 to true
    mockStore = setupStore({
      playerList: [
        {
          id: "1654A",
          name: "Alpha",
          score: 10,
          created: "11/21/2016",
          updated: "06/12/2018",
          isSelected: true,
        },
        {
          id: "1654B",
          name: "Beta",
          score: 4,
          created: "11/8/2016",
          updated: "11/9/2017",
          isSelected: false,
        },
      ],
    });
    wrapper = setupWrapper(mockStore);
    /* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "matchMultiLength"] }] */
    matchMultiLength(wrapper, [
      [".player.player-selected", 1],
      [".remove-player.remove-player-selected", 1],
    ]);
  });

  it("should handle removePlayer() dispatch", () => {
    wrapper.find(".remove-player").simulate("click");
    expect(remove).toHaveBeenCalledWith("1654A");
    expect(remove).toHaveBeenCalledTimes(1);
  });

  it("should handle selectPlayer() dispatch", () => {
    wrapper.find(".player-name").simulate("click");
    expect(select).toHaveBeenCalledWith("1654A");
    expect(select).toHaveBeenCalledTimes(1);
  });
});
