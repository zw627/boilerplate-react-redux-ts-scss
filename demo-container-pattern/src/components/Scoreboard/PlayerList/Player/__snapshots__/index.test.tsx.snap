// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard/Player compare to the last snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Player
    id="1654A"
    index={0}
    isSelected={false}
    name="Alpha"
    remove={[MockFunction]}
    select={[MockFunction]}
  >
    <div
      className="player"
    >
      <div
        className="remove-player"
        onClick={[Function]}
      >
        âœ•
      </div>
      <div
        className="player-name"
        onClick={[Function]}
      >
        <Connect(Crown)
          index={0}
        >
          <Crown
            dispatch={[Function]}
            hasCrown={true}
            index={0}
          >
            <svg
              className="crown has-crown"
              viewBox="0 0 44 35"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M26.7616 10.6207L21.8192 0L16.9973 10.5603C15.3699 14.1207 10.9096 15.2672 7.77534 12.9741L0 7.24138L6.56986 28.8448H37.0685L43.5781 7.72414L35.7425 13.0948C32.6685 15.2672 28.3288 14.0603 26.7616 10.6207Z"
                transform="translate(0 0.301727)"
              />
              <rect
                height="3.07759"
                transform="translate(6.56987 31.5603)"
                width="30.4986"
              />
            </svg>
          </Crown>
        </Connect(Crown)>
        <span>
          Alpha
        </span>
      </div>
      <Connect(Counter)
        index={0}
      >
        <Counter
          id="1654A"
          index={0}
          score={10}
          update={[Function]}
        >
          <div
            className="counter"
          >
            <button
              className="counter-action decrement"
              onClick={[Function]}
            >
              -
            </button>
            <div
              className="counter-score"
            >
              10
            </div>
            <button
              className="counter-action increment"
              onClick={[Function]}
            >
              +
            </button>
          </div>
        </Counter>
      </Connect(Counter)>
    </div>
  </Player>
</Provider>
`;
